File: 2023/domaci/44-matersky-programovaci-jazyk/volfmatous/main.cs
using System;
using System.Collections.Generic;
using System.Linq;

internal class Program
{
    static void Main(string[] args)
    {
        Dictionary<string, double> variables = new Dictionary<string, double>();

        int c = int.Parse(Console.ReadLine());

        List<string[]> linesSplit = new List<string[]>();
        List<int> linesNestLevels = new List<int>();

        for (int i = 0; i < c; i++)
        {
            string line = Console.ReadLine().Split("//")[0].TrimEnd();

            if (line == string.Empty)
            {
                continue;
            }

            int nestLevel = 0;
            while (line.Length > 2 && line.StartsWith("  "))
            {
                line = line.Substring(2);
                nestLevel++;
            }

            linesSplit.Add(line.Split(' '));
            linesNestLevels.Add(nestLevel);
        }

        List<double> output = new List<double>();

        bool repeating = false;
        int repeatStartLineIndex = 0;
        int nestLevelCurrent = 0;

        for (int i = 0; i < linesSplit.Count; i++)
        {
            string[] lineSplit = linesSplit[i];
            int nestLevel = linesNestLevels[i];

            if (nestLevel > nestLevelCurrent)
            {
                continue;
            }

            bool atLastLine = i >= linesSplit.Count - 1;

            if (repeating && (nestLevel < linesNestLevels[repeatStartLineIndex] + 1 || atLastLine))
            {
                repeating = false;
                i = repeatStartLineIndex - 1;
                if (!atLastLine)
                {
                    continue;
                }
            }
            else if (nestLevel < nestLevelCurrent)
            {
                nestLevelCurrent = nestLevel;
            }

            switch (lineSplit[0])
            {
                case "kdyz":
                {
                    if (lineSplit.Length < 5 || lineSplit[lineSplit.Length - 1] != "tak")
                    {
                        Console.WriteLine("error");
                        return;
                    }

                    (bool, bool) conditionResult = ConditionPass(lineSplit.Skip(1).SkipLast(1).ToArray(), variables);

                    if (!conditionResult.Item1)
                    {
                        Console.WriteLine("error");
                        return;
                    }

                    if (conditionResult.Item2)
                    {
                        nestLevelCurrent++;
                    }

                    break;
                }

                case "dokud":
                {
                    if (lineSplit.Length < 5 || lineSplit[lineSplit.Length - 1] != "tak")
                    {
                        Console.WriteLine("error");
                        return;
                    }

                    (bool, bool) conditionResult = ConditionPass(lineSplit.Skip(1).SkipLast(1).ToArray(), variables);

                    if (!conditionResult.Item1)
                    {
                        Console.WriteLine("error");
                        return;
                    }

                    if (conditionResult.Item2)
                    {
                        nestLevelCurrent++;
                        repeating = true;
                        repeatStartLineIndex = i;
                    }

                    break;
                }

                case "vypsat":
                {
                    (bool, double) expressionResult = ExpressionValue(lineSplit.Skip(1).ToArray(), variables);

                    if (!expressionResult.Item1)
                    {
                        Console.WriteLine("error");
                        return;
                    }

                    output.Add(expressionResult.Item2);
                    break;
                }

                default:
                {
                    string variableName = lineSplit[0];

                    if (char.ToLower(variableName[0]) < 'a' || char.ToLower(variableName[0]) > 'z'
                                                            || lineSplit.Length < 2 || lineSplit[1] != "je")
                    {
                        Console.WriteLine("error");
                        return;
                    }

                    switch (variableName)
                    {
                        case "je":
                        case "a":
                        case "bez":
                        case "kdyz":
                        case "tak":
                        case "rovno":
                        case "aspon":
                        case "nejvic":
                        case "nebo":
                        case "dokud":
                        case "vypsat":
                            Console.WriteLine("error");
                            return;
                    }

                    (bool, double) expressionResult = ExpressionValue(lineSplit.Skip(2).ToArray(), variables);

                    if (!expressionResult.Item1)
                    {
                        Console.WriteLine("error");
                        return;
                    }

                    variables[variableName] = expressionResult.Item2;
                    break;
                }
            }
        }

        foreach (double line in output)
        {
            Console.WriteLine(line);
        }
    }

    static (bool, double) ExpressionValue(string[] lineSplit, Dictionary<string, double> variables)
    {
        if (lineSplit.Length % 2 != 1)
        {
            return (false, 0);
        }

        bool expressionAdd = true;
        double expressionValue = 0;

        for (int w = 0; w < lineSplit.Length; w++)
        {
            if (w % 2 == 0)
            {
                double value;
                if (!double.TryParse(lineSplit[w], out value) && !variables.TryGetValue(lineSplit[w], out value))
                {
                    return (false, 0);
                }

                expressionValue += expressionAdd ? value : -value;
            }
            else
            {
                switch (lineSplit[w])
                {
                    case "a":
                        expressionAdd = true;
                        break;
                    case "bez":
                        expressionAdd = false;
                        break;
                    default:
                        return (false, 0);
                }
            }
        }

        return (true, expressionValue);
    }

    static (bool, bool) ConditionPass(string[] lineSplit, Dictionary<string, double> variables)
    {
        List<int> orIndexes = new List<int>();
        orIndexes.Add(-1);
        for (int i = 0; i < lineSplit.Length; i++)
        {
            if (lineSplit[i] == "nebo")
            {
                orIndexes.Add(i);
            }
        }

        orIndexes.Add(lineSplit.Length);

        bool conditionPass = false;

        for (int i = 0; i < orIndexes.Count - 1; i++)
        {
            int startIndex = orIndexes[i] + 1;
            int endIndex = orIndexes[i + 1] - 1;

            int comparatorIndex = -1;
            for (int c = startIndex; c <= endIndex; c++)
            {
                if (lineSplit[c] == "rovno" || lineSplit[c] == "aspon" || lineSplit[c] == "nejvic")
                {
                    comparatorIndex = c;
                    break;
                }
            }

            if (comparatorIndex == -1)
            {
                return (false, false);
            }

            string comparator = lineSplit[comparatorIndex];
            (bool, double) expression1Result = ExpressionValue(lineSplit
                    .Skip(startIndex)
                    .Take(comparatorIndex - startIndex)
                    .ToArray(),
                variables);
            (bool, double) expression2Result = ExpressionValue(lineSplit
                .Skip(comparatorIndex + 1)
                .Take(endIndex - comparatorIndex)
                .ToArray(), variables);

            if (!expression1Result.Item1 || !expression2Result.Item1)
            {
                return (false, false);
            }

            if (!conditionPass)
            {
                double expression1 = expression1Result.Item2;
                double expression2 = expression2Result.Item2;

                switch (comparator)
                {
                    case "rovno":
                        conditionPass = expression1 == expression2;
                        break;

                    case "aspon":
                        conditionPass = expression1 >= expression2;
                        break;

                    case "nejvic":
                        conditionPass = expression1 <= expression2;
                        break;
                }
            }
        }

        return (true, conditionPass);
    }
}

File: 2023/domaci/44-matersky-programovaci-jazyk/volfmatous/main.java
import java.text.DecimalFormat;
import java.util.Scanner;
import java.util.Map;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Map<String, Double> variables = new HashMap<>();

        int c = scanner.nextInt();
        scanner.nextLine();

        List<String[]> linesSplit = new ArrayList<>();
        List<Integer> linesNestLevels = new ArrayList<>();

        for (int i = 0; i < c; i++) {
            String line = scanner.nextLine().split("//")[0].stripTrailing();

            if (line.isEmpty()) {
                continue;
            }

            int nestLevel = 0;
            while (line.length() > 2 && line.startsWith("  ")) {
                line = line.substring(2);
                nestLevel++;
            }

            linesSplit.add(line.split(" "));
            linesNestLevels.add(nestLevel);
        }

        List<Double> output = new ArrayList<>();

        boolean repeating = false;
        int repeatStartLineIndex = 0;
        int nestLevelCurrent = 0;

        for (int i = 0; i < linesSplit.size(); i++) {
            String[] lineSplit = linesSplit.get(i);
            int nestLevel = linesNestLevels.get(i);

            if (nestLevel > nestLevelCurrent) {
                continue;
            }

            boolean atLastLine = i >= linesSplit.size() - 1;

            if (repeating && (nestLevel < linesNestLevels.get(repeatStartLineIndex) + 1 || atLastLine)) {
                repeating = false;
                i = repeatStartLineIndex - 1;
                if (!atLastLine) {
                    continue;
                }
            } else if (nestLevel < nestLevelCurrent) {
                nestLevelCurrent = nestLevel;
            }

            switch (lineSplit[0]) {
                case "kdyz": {
                    if (lineSplit.length < 5 || !lineSplit[lineSplit.length - 1].equals("tak")) {
                        System.out.println("error");
                        return;
                    }

                    boolean[] conditionResult = conditionPass(Arrays.copyOfRange(lineSplit, 1, lineSplit.length - 1), variables);

                    if (!conditionResult[0]) {
                        System.out.println("error");
                        return;
                    }

                    if (conditionResult[1]) {
                        nestLevelCurrent++;
                    }

                    break;
                }

                case "dokud": {
                    if (lineSplit.length < 5 || !lineSplit[lineSplit.length - 1].equals("tak")) {
                        System.out.println("error");
                        return;
                    }

                    boolean[] conditionResult = conditionPass(Arrays.copyOfRange(lineSplit, 1, lineSplit.length - 1), variables);

                    if (!conditionResult[0]) {
                        System.out.println("error");
                        return;
                    }

                    if (conditionResult[1]) {
                        nestLevelCurrent++;
                        repeating = true;
                        repeatStartLineIndex = i;
                    }

                    break;
                }

                case "vypsat": {
                    double expressionResult = expressionValue(Arrays.copyOfRange(lineSplit, 1, lineSplit.length), variables);

                    if (Double.isNaN(expressionResult)) {
                        System.out.println("error");
                        return;
                    }

                    output.add(expressionResult);
                    break;
                }

                default: {
                    String variableName = lineSplit[0];

                    if (Character.toLowerCase(variableName.charAt(0)) < 'a' || Character.toLowerCase(variableName.charAt(0)) > 'z'
                            || lineSplit.length < 2 || !lineSplit[1].equals("je")) {
                        System.out.println("error");
                        return;
                    }

                    switch (variableName) {
                        case "je":
                        case "a":
                        case "bez":
                        case "kdyz":
                        case "tak":
                        case "rovno":
                        case "aspon":
                        case "nejvic":
                        case "nebo":
                        case "dokud":
                        case "vypsat":
                            System.out.println("error");
                            return;
                    }

                    double expressionResult = expressionValue(Arrays.copyOfRange(lineSplit, 2, lineSplit.length), variables);

                    if (Double.isNaN(expressionResult)) {
                        System.out.println("error");
                        return;
                    }

                    variables.put(variableName, expressionResult);
                    break;
                }
            }
        }

        for (double line : output) {
            System.out.println(new DecimalFormat("#.################").format(line));
        }
    }

    static boolean[] conditionPass(String[] lineSplit, Map<String, Double> variables) {
        List<Integer> orIndexes = new ArrayList<>();
        orIndexes.add(-1);
        for (int i = 0; i < lineSplit.length; i++) {
            if (lineSplit[i].equals("nebo")) {
                orIndexes.add(i);
            }
        }

        orIndexes.add(lineSplit.length);

        boolean conditionPass = false;

        for (int i = 0; i < orIndexes.size() - 1; i++) {
            int startIndex = orIndexes.get(i) + 1;
            int endIndex = orIndexes.get(i + 1) - 1;

            int comparatorIndex = -1;
            for (int c = startIndex; c <= endIndex; c++) {
                if (lineSplit[c].equals("rovno") || lineSplit[c].equals("aspon") || lineSplit[c].equals("nejvic")) {
                    comparatorIndex = c;
                    break;
                }
            }

            if (comparatorIndex == -1) {
                return new boolean[]{false, false};
            }

            String comparator = lineSplit[comparatorIndex];
            double expression1Result = expressionValue(Arrays.copyOfRange(lineSplit, startIndex, comparatorIndex), variables);
            double expression2Result = expressionValue(Arrays.copyOfRange(lineSplit, comparatorIndex + 1, endIndex + 1), variables);

            if (Double.isNaN(expression1Result) || Double.isNaN(expression2Result)) {
                return new boolean[]{false, false};
            }

            if (!conditionPass) {
                double expression1 = expression1Result;
                double expression2 = expression2Result;

                switch (comparator) {
                    case "rovno":
                        conditionPass = expression1 == expression2;
                        break;

                    case "aspon":
                        conditionPass = expression1 >= expression2;
                        break;

                    case "nejvic":
                        conditionPass = expression1 <= expression2;
                        break;
                }
            }
        }

        return new boolean[]{true, conditionPass};
    }

    static double expressionValue(String[] lineSplit, Map<String, Double> variables) {
        if (lineSplit.length % 2 != 1) {
            return Double.NaN;
        }

        boolean expressionAdd = true;
        double expressionValue = 0;

        for (int w = 0; w < lineSplit.length; w++) {
            if (w % 2 == 0) {
                double value;
                try {
                    value = Double.parseDouble(lineSplit[w]);
                } catch (NumberFormatException e) {
                    if (!variables.containsKey(lineSplit[w])) {
                        return Double.NaN;
                    }

                    value = variables.get(lineSplit[w]);
                }

                expressionValue += expressionAdd ? value : -value;
            } else {
                switch (lineSplit[w]) {
                    case "a":
                        expressionAdd = true;
                        break;
                    case "bez":
                        expressionAdd = false;
                        break;
                    default:
                        return Double.NaN;
                }
            }
        }

        return expressionValue;
    }
}
